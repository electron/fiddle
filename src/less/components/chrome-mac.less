.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 24px;
    font-size: 1.4rem;
    line-height: 15px;

    small,
    p,
    span {
      user-select: none;
    }
  }

  .big_sur_or_later {
    height: 30px !important;
  }
}
