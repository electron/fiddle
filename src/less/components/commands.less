@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-3;
  -webkit-app-region: drag;
}

.disabled-menu-tooltip {
  display: flex;
  width: 100%;

  .bp3-popover-target {
    width: 100%;
  }
}

.commands.is-mac {
  padding:
    10px
    calc(100vw - (env(titlebar-area-x, 0) + env(titlebar-area-width, 0)))
    10px
    env(titlebar-area-x, 0);
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding:
    10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
    white-space: nowrap;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  fieldset {
    border: 0;
    padding: 0;
    margin: 0;
    min-width: 0;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      .bp3-button-group {
        margin-left: 5px;
      }
    }
  }

  .bp3-control-group {
    margin-right: 0.5rem;
  }

  .address-bar {
    width: 420px;
    transition: width 0.4s;

    input,
    span {
      -webkit-app-region: no-drag;
    }

    &.empty {
      width: 300px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }

  & > .title {
    align-self: stretch;
    flex-shrink: 0;

    flex-flow: column nowrap;
    justify-content: center;
    user-select: none;

    margin: 0 10px;
  }
}
