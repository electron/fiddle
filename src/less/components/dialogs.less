@import "../variables.less";

.dialogs {
  z-index: 10;
  position: fixed;
  height: 100vh;
  width: 100vw;
  pointer-events: none;

  * {
    pointer-events: all;
  }
}

.dialog {
  background-color: @background-3;
  border: @border;
  color: @text-color-1;
  border-radius: 6px;
  padding: 10px;
  z-index: 100;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  animation: slide-up 0.4s ease;

  p {
    margin-bottom: 1.25rem;
    font-size: 1.25rem;
  }

  code {
    font-size: 12px;
    background: rgba(255, 255, 255, 0.1);
    padding: 2px;
    line-height: 12px;
  }

  &-buttons {
    margin-top: 1.25rem;
  }

  &.centered {
    position: absolute;
    width: 300px;
    left: calc(50% - (300px / 2));
    top: 100px;
  }

  &.arrow::after,
  &.arrow::before {
    border: solid transparent;
    content: " ";
    height: 0;
    width: 0;
    position: absolute;
    pointer-events: none;
  }

  &.arrow::after {
    border-color: rgba(136, 183, 213, 0);
    border-width: 9px;
  }

  &.arrow::before {
    border-color: rgba(194, 225, 245, 0);
    border-width: 10px;
  }

  &.arrow-top::after,
  &.arrow-top::before {
    bottom: 100%;
    left: 50%;
  }

  &.arrow-top::after {
    border-bottom-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-top::before {
    border-bottom-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-bottom::after,
  &.arrow-bottom::before {
    top: 100%;
    left: 50%;
  }

  &.arrow-bottom::after {
    border-top-color: @background-3;
    margin-left: -9px;
  }

  &.arrow-bottom::before {
    border-top-color: @border-color-1;
    margin-left: -10px;
  }

  &.arrow-left::after,
  &.arrow-left::before {
    right: 100%;
    top: 20px;
  }

  &.arrow-left::after {
    border-right-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-left::before {
    border-right-color: @border-color-1;
    margin-top: -10px;
  }

  &.arrow-right::after,
  &.arrow-right::before {
    left: 100%;
    top: 20px;
  }

  &.arrow-right::after {
    border-left-color: @background-3;
    margin-top: -9px;
  }

  &.arrow-right::before {
    border-left-color: @border-color-1;
    margin-top: -10px;
  }

  button {
    color: #1e2527;
    float: right;
    margin-left: 9px;
    margin-bottom: 0;
  }

  a {
    cursor: pointer;
  }

  h4 {
    margin-bottom: 0;
  }

  p:last-of-type {
    margin-bottom: 0;
  }

  input {
    color: rgba(159, 234, 249, 0.5);
    height: 28px;
    margin-top: 10px;
    background-color: @background-1;
    border: 1px solid @border-color-2;
    outline: 0 solid transparent;

    &:focus {
      border-color: @foreground-2;
    }
  }
}

.dialogDrop {
  position: fixed;
  z-index: 100;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  background: black;
  transition: opacity 0.4s ease-in-out;
  pointer-events: all;
  opacity: 0.5;
}

.dialog-add-version {
  .bp3-file-input {
    width: 100%;
  }

  .bp3-callout {
    margin-top: 1rem;
  }
}

select {
  font-size: 12px;
  height: 28px;
  background-color: white;
  line-height: 15px;
}

@keyframes slide-up {
  0% {
    opacity: 0;
    transform: translateY(20px);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
