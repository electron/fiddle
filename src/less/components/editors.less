.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 5px 0 5px;

    .mosaic-controls {
      transform: scale(0.75);
      transform-origin: right;
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

// TODO: support new file
// update if list of Editor ID changes
@editor-ids: main\.js, renderer\.js, index\.html, preload\.js, styles\.css;

// takes each editor ID and increase its z-index if the parent Mosaic root
// has focused__id class for that specific id.
each(@editor-ids, {
  .focused__@{value}.mosaic .mosaic-window.@{value} {
    z-index: 2;
  }
});

.mosaic-window {
  z-index: 1;
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
