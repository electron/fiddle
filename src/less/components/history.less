.revision-list {
  max-height: 400px;
  overflow-y: auto;
}

.revision-list ul {
  list-style-type: none;
  padding: 0;
  margin: 0;
}

.revision-item {
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 3px;
  cursor: pointer;
  transition: background-color 0.2s ease;
  border-left: 3px solid #106ba3;
}

.revision-item:hover {
  background-color: rgba(167, 182, 194, 0.3);
}

.revision-content {
  display: flex;
  flex-direction: column;
}

.revision-icon {
  margin-right: 8px;
}

.sha-label {
  font-size: 12px;
  color: #738694;
  margin-left: 10px;
  font-family: monospace;
}

.revision-details {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-top: 5px;
}

.revision-date {
  color: #738694;
  font-size: 12px;
}

.revision-changes {
  display: flex;
  gap: 5px;
}

.history-loading {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
}
