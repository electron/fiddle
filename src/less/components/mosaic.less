@import '../variables.less';

.fiddle {
  .mosaic-window-toolbar {
    background: var(--background-1);
    box-shadow: inset 0 0 100px 100px rgba(53, 53, 53, 0.1);
    border-top: 1px solid rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-toolbar.draggable:hover {
    box-shadow: inset 0 0 100px 100px rgba(255, 255, 255, 0.1);
  }

  .mosaic-window-title {
    color: var(--text-color-3);
  }

  button.mosaic-default-control {
    padding: 0;
    height: 20px;
    width: 20px;
    margin: 5px;
    background-color: #314246;
    border: 0.1rem solid #222b35;
  }

  button.mosaic-default-control:hover {
    background-color: #131d1f;
  }

  .mosaic-window-body {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: stretch;
    align-content: stretch;
    overflow: visible;
    z-index: 4;
    background-color: @background-4;
  }

  .mosaic.mosaic-blueprint-theme {
    background: #bdbdbd15;
  }

  .editorContainer,
  .monaco-editor {
    width: 100%;
    height: 100%;
  }

  .editorContainer {
    background: @background-1;
  }
}
