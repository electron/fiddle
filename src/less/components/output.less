@import '../variables.less';

.output {
  font-size: 12px;
  font-family: @fonts-common;
  font-weight: 600;
  top: 10vh;
  padding: 0;
  height: 100%;
  -webkit-app-region: no-drag;

  .monaco-editor {
    color: var(--text-color-1);
    box-shadow: inset 0 0 100px 100px rgba(0, 0, 0, 0.15);
    border-top: @border;

    .margin,
    .monaco-editor-background {
      background: var(--background-2);
    }
  }
}
