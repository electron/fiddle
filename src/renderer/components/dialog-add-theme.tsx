import * as React from 'react';

import * as path from 'path';

import { Button, Dialog, FileInput } from '@blueprintjs/core';
import { shell } from 'electron';
import * as fs from 'fs-extra';
import { observer } from 'mobx-react';
import * as MonacoType from 'monaco-editor';

import { AppState } from '../state';
import { THEMES_PATH, getTheme } from '../themes';
import { LoadedFiddleTheme, defaultDark } from '../themes-defaults';

interface AddThemeDialogProps {
  appState: AppState;
}

interface AddThemeDialogState {
  file?: File;
}

/**
 * The "add monaco theme" dialog allows users to add custom editor themes.
 *
 * @class AddThemeDialog
 * @extends {React.Component<AddThemeDialogProps, AddThemeDialogState>}
 */
export const AddThemeDialog = observer(
  class AddThemeDialog extends React.Component<
    AddThemeDialogProps,
    AddThemeDialogState
  > {
    public resetState = { file: undefined };

    constructor(props: AddThemeDialogProps) {
      super(props);
      this.state = this.resetState;

      this.onSubmit = this.onSubmit.bind(this);
      this.onClose = this.onClose.bind(this);
      this.onChangeFile = this.onChangeFile.bind(this);
      this.reset = this.reset.bind(this);
    }

    /**
     * Handles a change of the file input.
     *
     * @param {React.ChangeEvent<HTMLInputElement>} event
     */
    public async onChangeFile(event: React.FormEvent<HTMLInputElement>) {
      const { files } = event.target as any;
      const file = files?.[0];

      this.setState({ file });
    }

    /**
     * Handles the submission of the dialog.
     *
     * @returns {Promise<void>}
     */
    public async onSubmit(): Promise<void> {
      const { file } = this.state;
      const { appState } = this.props;

      const defaultTheme = !!appState.theme
        ? await getTheme(appState.theme)
        : defaultDark;

      if (!file) return;

      try {
        const editor = fs.readJSONSync(file.path);
        if (!editor.base && !editor.rules)
          throw Error('File does not match specifications'); // has to have these attributes
        defaultTheme.editor = editor as Partial<MonacoType.editor.IStandaloneThemeData>;
        const newTheme = defaultTheme;
        const name = editor.name ? editor.name : file.name;
        await this.createNewThemeFromMonaco(name, newTheme);
      } catch (error) {
        appState.showErrorDialog(`${error}, please pick a different file.`);
        return;
      }

      this.onClose();
      return;
    }

    public async createNewThemeFromMonaco(
      name: string,
      newTheme: LoadedFiddleTheme,
    ): Promise<void> {
      if (!name) {
        throw new Error(`Filename ${name} not found`);
      }

      const themePath = path.join(THEMES_PATH, `${name}`);

      await fs.outputJSON(
        themePath,
        {
          ...newTheme,
          name,
        },
        { spaces: 2 },
      );

      this.props.appState.setTheme(themePath);
      shell.showItemInFolder(themePath);
    }

    get buttons() {
      const canSubmit = !!this.state.file;

      return [
        <Button
          icon="add"
          key="submit"
          disabled={!canSubmit}
          onClick={this.onSubmit}
          text="Add"
        />,
        <Button
          icon="cross"
          key="cancel"
          onClick={this.onClose}
          text="Cancel"
        />,
      ];
    }

    public onClose() {
      this.props.appState.isThemeDialogShowing = false;
      this.reset();
    }

    public render() {
      const { isThemeDialogShowing } = this.props.appState;
      const inputProps = { accept: '.json' };
      const { file } = this.state;

      const text = file && file.path ? file.path : `Select the Monaco file...`;
      return (
        <Dialog
          isOpen={isThemeDialogShowing}
          onClose={this.onClose}
          title="Add theme"
          className="dialog-add-version"
        >
          <div className="bp3-dialog-body">
            <FileInput
              onInputChange={this.onChangeFile}
              inputProps={inputProps as any}
              text={text}
            />
            <br />
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">{this.buttons}</div>
          </div>
        </Dialog>
      );
    }

    /**
     * Reset this component's state
     */
    private reset(): void {
      this.setState(this.resetState);
      return;
    }
  },
);
