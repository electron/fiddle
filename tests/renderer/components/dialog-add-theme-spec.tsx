import * as React from 'react';

import * as path from 'path';

import { shell } from 'electron';
import { shallow } from 'enzyme';
import * as fs from 'fs-extra';

import { AddThemeDialog } from '../../../src/renderer/components/dialog-add-theme';
import {
  LoadedFiddleTheme,
  defaultLight,
} from '../../../src/renderer/themes-defaults';
import { StateMock } from '../../mocks/mocks';
import { overridePlatform, resetPlatform } from '../../utils';

jest.mock('../../../src/renderer/ipc');

describe('AddThemeDialog component', () => {
  let store: StateMock;

  beforeAll(() => {
    // We render the buttons different depending on the
    // platform, so let' have a uniform platform for unit tests
    overridePlatform('darwin');
  });

  afterAll(() => {
    resetPlatform();
  });

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);
  });

  it('renders', () => {
    const wrapper = shallow(<AddThemeDialog appState={store as any} />);

    wrapper.setState({
      file: '/test/file',
    });

    expect(wrapper).toMatchSnapshot();
  });

  describe('createNewThemeFromMonaco()', () => {
    it('handles invalid input', async () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      try {
        await instance.createNewThemeFromMonaco('', {} as LoadedFiddleTheme);
      } catch (err) {
        expect(err.message).toEqual(`Filename  not found`);
        expect(fs.outputJSON).toHaveBeenCalledTimes(0);
        expect(store.setTheme).toHaveBeenCalledTimes(0);
        expect(shell.showItemInFolder).toHaveBeenCalledTimes(0);
      }
    });

    it('handles valid input', async () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);
      const instance: any = wrapper.instance() as any;
      wrapper.setState({ file: '/test/file' });

      (fs.outputJSON as jest.Mock).mockResolvedValue({});

      await instance.createNewThemeFromMonaco('testingLight', defaultLight);

      expect(fs.outputJSON).toHaveBeenCalled();

      const args = (fs.outputJSON as jest.Mock).mock.calls[0];
      expect(args[0].includes(`.electron-fiddle`)).toBe(true);
      expect(args[1].name).toEqual('testingLight');
      expect(args[1].common).toBeDefined();
      expect(args[1].file).toBeDefined();

      const themePath = path.join(
        '~',
        '.electron-fiddle',
        'themes',
        'testingLight',
      );
      expect(store.setTheme).toHaveBeenCalledWith(themePath);
      expect(shell.showItemInFolder).toHaveBeenCalledWith(themePath);
    });
  });

  describe('onSubmit()', () => {
    it('does nothing if there is no file currently set', async () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      instance.createNewThemeFromMonaco = jest.fn();
      instance.onClose = jest.fn();

      await instance.onSubmit();

      expect(fs.readJSONSync).toHaveBeenCalledTimes(0);
      expect(instance.createNewThemeFromMonaco).toHaveBeenCalledTimes(0);
      expect(instance.onClose).toHaveBeenCalledTimes(0);
    });

    it('loads a theme if a file is currently set', async () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      wrapper.setState({ file: '/test/file' });

      (fs.readJSONSync as jest.Mock).mockReturnValue(defaultLight.editor);

      instance.createNewThemeFromMonaco = jest.fn();
      instance.onClose = jest.fn();

      await instance.onSubmit();

      expect(fs.readJSONSync).toHaveBeenCalledTimes(1);
      expect(instance.createNewThemeFromMonaco).toHaveBeenCalledTimes(1);
      expect(instance.onClose).toHaveBeenCalledTimes(1);
    });

    it('shows an error dialog for a malformed theme', async () => {
      store.showErrorDialog = jest.fn().mockResolvedValueOnce(true);
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      wrapper.setState({ file: '/test/file' });

      (fs.readJSONSync as jest.Mock).mockReturnValue({});

      instance.onClose = jest.fn();

      await instance.onSubmit();

      expect(fs.readJSONSync).toHaveBeenCalledTimes(1);
      expect(store.showErrorDialog).toHaveBeenCalledWith(
        expect.stringMatching(/file does not match specifications/i),
      );
    });
  });

  describe('onChangeFile()', () => {
    it('handles valid input', () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);

      const files = ['one', 'two'];
      (wrapper.instance() as any).onChangeFile({
        target: { files },
      });
      expect(wrapper.state('file')).toBe(files[0]);
    });

    it('handles no input', () => {
      const wrapper = shallow(<AddThemeDialog appState={store as any} />);

      (wrapper.instance() as any).onChangeFile({
        target: { files: null },
      });
      expect(wrapper.state('file')).toBeUndefined();
    });
  });
});
