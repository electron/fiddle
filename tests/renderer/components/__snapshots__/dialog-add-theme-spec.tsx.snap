// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddThemeDialog component renders 1`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  className="dialog-add-version"
  isOpen={false}
  onClose={[Function]}
  title="Add theme"
>
  <div
    className="bp3-dialog-body"
  >
    <Blueprint3.FileInput
      hasSelection={false}
      inputProps={
        Object {
          "accept": ".json",
        }
      }
      onInputChange={[Function]}
      text="Select the Monaco file..."
    />
    <br />
  </div>
  <div
    className="bp3-dialog-footer"
  >
    <div
      className="bp3-dialog-footer-actions"
    >
      <Blueprint3.Button
        disabled={false}
        icon="add"
        key="submit"
        onClick={[Function]}
        text="Add"
      />
      <Blueprint3.Button
        icon="cross"
        key="cancel"
        onClick={[Function]}
        text="Cancel"
      />
    </div>
  </div>
</Blueprint3.Dialog>
`;
